### ISSUE TITLE
Application template with Login & Tabs

### ISSUE THREAD

**Issue by 0419** [leeoniya/domvm#69]
Hello

Would it be possible to create an example of how application should be structured which has login page and then few tabs inside it. I tried to look at Todo & ThreadIt, both are structured very different ways.

**Comment by leeoniya** [comment]
Sure, i can put together one potential starting point. Do you plan to make your backend just a rest/json api and do the rest on the client?

**Comment by 0419** [comment]
yes, this is for intranet applications. At moment everything is in plain JavaScript, no newer syntax yet. We have been making use of https://github.com/mitsuhiko/classy and following is MVC our framework.  

```
(function(global) {

	/**
	* Extend this class for every REST API on server
	*
	*/
	var Model = Class.$extend({
		__init__ : function(url) {
			this.url = url;
		},

		save: function(data, onSuccess, onFailure) {
			var server = global.App.Server;
			server.form_post(this.url, data, onSuccess, onFailure);
		},

		destroy: function(data, onSuccess, onFailure) {
			var server = global.App.Server;
			//server.delete(this.url, data, onSuccess, onFailure);
		},

		fetch: function() {
			var self = this;

			App.Server.get(self.url, function(data) {
				self.data = data;
			})
		}
	});

	var eventSplitter = /^(\S+)\s*(.*)$/;

	/**
	* Use this for all views
	*
	*/
	var View = Class.$extend({
		__init__ : function() {
			this.cid = _.uniqueId('view');
		},

		html: function(tpl, context) {
			var tmpl = global.Utils.Tpl.get(tpl);
			var html = tmpl(context);

			return html;
		},

		template: function(tpl, context) {
			var html = this.html(tpl, context);
			this.$el = $(html);

			return this.$el;
		},

		e: function(sel) {
			return this.$el.find(sel);
		}
	});

	// assign the variable to window
	global.Model = Model;
	global.View  = View;

})(this);
```

**Comment by leeoniya** [comment]
In the interest of creating a reusable scaffold that can be added into the repo, let's figure out what we expect the server to provide for us.

The server:
	- provides a REST/JSON API that returns correct http status codes
	- auth/sessions may expire and return a 401, should notify of sess expire, app should redirect to some `/auth` route for re-login
	- any request can fail (4\*\*/5\*\* status codes) and needs to be handled at some general level

Hash or history routing?

Hash routing is more flexible because it requires the least amount of server config. The server only ever gets the initial request (or F5 refresh) for all pages to single root app url, the sub-section nav is then re-requested via ajax and rendered once the JSON state is decoded:

- `https://admin.mydomain.net/#/users` - server gets request to `/`
- `https://mydomain.net/admin/#/users` - server gets request to `/admin`

History routing requires the server to be configured to know the app root, and the client router must also be configured to know where app root starts/ends.

- `https://admin.mydomain.net/users` - server gets request to `/users`       but must know to return `/` root index and allow the client re-request `/users` portion
- `https://mydomain.net/admin/users` - server gets request to `/admin/users` but must know to return `/admin` root index and allow the client re-request `/users` portion

Should app state dumped into the initial request's markup (in script tags) or does the app do an additional ajax request on init (e.g. `GET /state`) to fetch app state (logged-in status/user, etc) as it would for all followup navigation/polling? The ajax approach adds an extra http request but is more uniform and probably a bit easier to implement. In addition, should `GET /state` requests fetch the full application state or should they fetch only the "active" views' states? Should these be passed as a query params?: `GET /state`, `GET /state?views=users,notifs`.

Once we settle some of these possibilities, it will be possible to create a structure that handles the intended setup.

**Comment by 0419** [comment]
wow, that's good and all points are valid. It would be good idea to keep server stateless as much as possible. Hash based routing is good. At present, there is no routing at all. So users cannot bookmark a page or receive a URL in outlook. 

Following are some other nice to have features list in the command server call interceptor

At moment I have a JS object "Server" which all other JS objects use to make server calls. It takes care of adding AUTH header to each request if available. It also takes care of capturing failures and displaying alert to user. The success wrapper method checks if response is result of some download request, then accordingly changes download iframe src attribute to it, so download starts. Apart from regular ajax calls, there is need for file upload too, which makes a multipart request.

```		_success: function(data, textStatus, jqXHR, onSuccess) {
			if (_.isFunction(onSuccess)) {
				onSuccess(data, textStatus, jqXHR);
			} else if (_.isString(data)) {
				var n = data.indexOf("download:");
				if (n == 0) {
					var url = encodeURIComponent(config.AUTH) + '/' + encodeURIComponent(data.substring(9));
					Server.download(url);

					Utils.callIf(app.main.activeTab.downloadComplete);
				} else {
					console.log('unknown response 1 ' + data);
				}
			} else {
				console.log('unknown response 2 ' + data);
			}
		},

		_failure: function(jqXHR, textStatus, errorThrown, onFailure) {
			// don't modify 'errorThrown'
			var msg = errorThrown;

			if (_.isString(jqXHR.responseText)) {
				msg = errorThrown + ':' + jqXHR.responseText;
			}

			if (errorThrown === 'Server Error') {
				msg = jqXHR.responseText;
			}

			if (errorThrown === 'timeout') {
				msg = 'Your request has timed out, please consider narrowing your search';
			}

			console.log(msg);
			//console.log('status code :: ' + jqXHR.status);

			if (app.main) {
				if (app.main.activeTab) {
					var tmpActiveTab = app.main.activeTab;

					if (jqXHR.status === 403 || errorThrown === 'Forbidden') {
						tmpActiveTab.isForbidden = true;
					}

					if (tmpActiveTab.initialized) {
						if (_.isFunction(tmpActiveTab.showAlert)) {
							tmpActiveTab.showAlert(msg);
						}
					}
					else {
						tmpActiveTab.errorBeforeInitialized = true;
						tmpActiveTab.errorMsg = msg;
					}
				}
			}

			if (_.isFunction(onFailure)) {
				onFailure(jqXHR, textStatus, errorThrown);
			}
		},

		// download logic
		download : function(url) {
			var iframe;
			iframe = document.getElementById("hiddenDownloader");

			if (iframe === null) {
				iframe = document.createElement('iframe');
				iframe.id = "hiddenDownloader";
				iframe.style.display = 'none';
				document.body.appendChild(iframe);
			}

			iframe.src = 'api/download/excel/' + url;
		}
```

**Comment by leeoniya** [comment]
> Apart from regular ajax calls, there is need for file upload too, which makes a multipart request.

If this is for an internal app and you don't need to support old browsers, you could do multi-part file uploads using either Fetch or FormData+XMLHttpRequest [1] [2].

Regarding som general scaffold, the basic order of operations is:

- request to server, get back html, app.js
- get `/state` json either via ajax or from the html if it was already dumped there on initial request.
- init the router which in turn parses the current route and mounts the app, passing it the state and the active "view" it should show (which basically becomes part of the state). pass the router itself into the app view during app mounting so it can be used to generate hrefs for templates and can be used for triggering route changes from handlers, etc.

i'll put something up in a day or two, but it'll be pretty similar to https://github.com/leeoniya/domvm/blob/1.x-dev/demos/threaditjs/app.js#L3-L49 - `ThreaditApp` by itself is mostly just the logic/state/data fetching code (the model) which gets passed into `ThreaditView` for rendering.

[1] http://zinoui.com/blog/ajax-file-upload
[2] https://developer.mozilla.org/en-US/docs/Using_files_from_web_applications

**Comment by leeoniya** [comment]
@0419 just an update.

I'm fixing a couple minor bugs i found in the router while making a demo impl of AndVote [1]. I'm keeping this issue top-of-mind to make sure there's a decent routing scaffold to show. I'll be adding some config options for setting up default HTTP error handlers to DRY out some of the initial scaffolds I've tried doing so far. Stay tuned...

[1] https://www.reddit.com/r/javascript/comments/4hurvs/live_coding_a_javascript_poll_using_react_express/

**Comment by 0419** [comment]
Hi, @leeoniya any updates on this

**Comment by leeoniya** [comment]
@0419

I have some code to push but I'm in Hawaii for my wedding till June 1, so prolly not until then ;)

**Comment by lawrence-dol** [comment]
> I'm in Hawaii for my wedding till June 1

Hey, congratulations man! All the best for the future.

**Comment by leeoniya** [comment]
Thanks guys :P

I'm catching up on stuff, but expect some updates soon.

**Comment by leeoniya** [comment]
I'm dedicating some time this weekend to getting this done. I've been back-and-forth on the architecture since even for a skeleton there are a surprising amount of opinions that end up getting baked in. Here are some guiding principals and the general design / flow.

One of the main questions was how to deliver an app that was secure by default. Typically, SPAs are delivered in a single bundle for the whole app, but what if we dont want to expose all the code to everyone? What if depending on the user's role/permissions they only have access to specific modules? This means that an Auth module must be the gatekeeper to which modules' code is actually sent to the client. This in turn implies that modules must be able to load dynamically, adding routes and augmenting nav menus on-demand.

If the a user is not logged in, only the code for the Auth module/view is ever sent to the client.

1. initial https request
2. server returns
  - html skeleton
  - App.js skeleton with base layout/view
  - Auth.js module & state (if session exists)
  - Any additional modules & states for which the logged in user has permissions
3. App.js then..
  - loads Auth module plus any others
  - Initializes router & assembles nav from all provided module configs
  - invokes router to render module/view that corresponds to url

I'm currently working out what the per-module config should look like and if they can nest recursively, declaring sub-routes and sub-nav.

Each module's state will need some toggleable "active" flag that short-circuits redraw() calls,  assuming we decide to retain DOM of inactive modules and simply toggle them via display: none. Alternatively we can just swap modules and re-create/replace the dom upon switching. Again, opinions are plentiful but depend a lot on how heavy each module is expected to be, which makes any generic solution suboptimal for numerous cases.

**Comment by franzliedke** [comment]
Not sure whether this belongs here, but...
How exactly does splitting up code for logged-in users vs. anonymous users make the app more secure, if it's trivial to sign up for an account?

**Comment by leeoniya** [comment]
@franzliedke 

> if it's trivial to sign up for an account?

for non-public admin backends (no anon users), an account may need to be created by HR with some specific permissions/role (accounting, off-shore contractors) that limit your module access. i work for a company where this is the case.

**Comment by grumpi** [comment]
@leeoniya The idea of loading parts of the app based on whether the authenticated user can use them is interesting, but I think it's also a bit of a niche use case.

In many cases it seems to me that what needs to be protected at all cost is data - not code. Though, I can see that there are cases where the code itself is a secret that should not be exposed to unauthorized users.

In the case where the code is secret, it seems that you can't use a simple static file server to serve the protected modules - they must all be served by a server running under TLS that explicitly checks authorization. The simplest way I can think of to achieve different code being served based on authorization is:

* have a very basic login page that isn't part of the SPA itself (either that, or have a minimal "login-only" build of the SPA) and a server that implements traditional cookie/csrf auth
* have different SPA builds depending on the different authorization levels. `require` only what is needed into each build while trying to reuse code between the different builds. Likely involves a central file with glue code for each build.
* after login, force a reload/redirect that makes the user's browser send a request to the server. Have the server render a HTML skeleton that links the appropriate resources for the SPA build the user needs to be served (based on their session cookie)
* confidential SPA builds are served from the same domain as the login server. This implies that any request for them includes the session cookie of the authorized user. In turn, this means that the server can check authorization before serving the `js`-files.

This way you won't have to deal with live-loading of code while the SPA runs. When privileges change, make your users reload the page from the server who will serve a HTML skeleton with the appropriate SPA build.

I know this is a pretty low-tech approach and page-reloads aren't really considered a sophisticated UX, but considering that the problem is also very niche, maybe a lightweight approach makes sense?

Any case, I feel it makes a lot of sense to talk about different use cases and ideas what can be done. I currently only have a very very basic login example as part of https://github.com/grumpi/domvm-isomorphic-example. There, the app is served as a single `js` file and login/logout is just mocked.

**Comment by lawrence-dol** [comment]
I completely concur with @grumpi. Serving different coffee based on authorization is complex and error prone. And it's revisited for every client.

It is far simpler to secure the APIs at the edge. Then you automatically have full inter-op for all clients.

**Comment by leeoniya** [comment]
@0419 

i've put up an admin example [1].

Some things to note...

If it seems overly complicated, it's because I develop API-first. Anything that you can do with the UI, you should be able to do via an API. Because of this I split up my modules and views which results in more code, but also better concern separation.

There are some other design choices [rather than domvm restrictions] made here, such as all modules exposing a single view and being pre-initialized during init (including view construction). This example loads the entire SPA at once including the auth page.

You could also choose to access `router` and `app` globally (since they're singletons anyhow) and not have to pass them down into the views every time which can be a hassle.

I tried to leave some comments in the source so hopefully they help.

[1] http://leeoniya.github.io/domvm/demos/admin/
[2] https://github.com/leeoniya/domvm/blob/1.x-dev/demos/admin/index.html#L163

**Comment by 0419** [comment]
Excellent, thank you very much. I'll probably use it next week and let you know how it goes.

**Comment by 0419** [comment]
Closing this issue for now
